src/app/dataset-view/dataset.service.ts

```diff
@@ -303,66 +230,140 @@ export class AppDatasetService {
                     );
                 this.searchDatasetInfoChanges(dataset);
 
-                this.updateDatasetTree(data);
+                const lineageResponse: DatasetLineageNode = this.lineageResponseFromRawQuery(data);
+                this.updatelineageGraph(lineageResponse);
             });
     }
 
-    private updateDatasetTree(lineage: GetDatasetLineageQuery) {
-        const tree: DatasetKindInterface[][] = [];
-        const origin = lineage.datasets.byOwnerAndName as Dataset;
-        this.updateDatasetTreeRec(tree, origin);
-        this.datasetTreeChange(tree, origin);
+    private lineageResponseFromRawQuery(data: GetDatasetLineageQuery): DatasetLineageNode {
+        if (isNil(data.datasets.byOwnerAndName)) {
+            throw new Error("Dataset not resolved by ID");
+        }
+        const originDatasetBasics = data.datasets.byOwnerAndName as DatasetBasicsFragment;
+        const originMetadata = data.datasets.byOwnerAndName.metadata;
+
+        return {
+            basics: originDatasetBasics,
+            downstreamDependencies: originMetadata.currentDownstreamDependencies.map(
+                (downDependency) => {
+                    return {
+                        basics: downDependency as DatasetBasicsFragment,
+                        downstreamDependencies: downDependency.metadata.currentDownstreamDependencies.map(
+                            (downDependency2) => {
+                                return {
+                                    basics: downDependency2 as DatasetBasicsFragment,
+                                    downstreamDependencies: downDependency2.metadata.currentDownstreamDependencies.map(
+                                        (downDependency3) => {
+                                            return {
+                                                basics: downDependency3 as DatasetBasicsFragment,
+                                                downstreamDependencies: downDependency3.metadata.currentDownstreamDependencies.map(
+                                                    (downDependency4) => {
+                                                        return {
+                                                            basics: downDependency4 as DatasetBasicsFragment,
+                                                            downstreamDependencies: downDependency4.metadata.currentDownstreamDependencies.map(
+                                                                (downDependency5) => {
+                                                                    return {
+                                                                        basics: downDependency5 as DatasetBasicsFragment,
```

Comment:
This probably could be expressed more concisely through recursion

Reply:
The whole point of this is that there is no way to do it in a completely type-safe manner. First of all, there are no explicit types of parts, those generated union/intersection types do not have any name. Second each level is a different data type. So, we cannot really express it recursively with how the base GraphQL query is defined.